Feature/credit card validator

onurersel opened the pull request:
### Summary
Added credit card validation.

onurersel committed:
Added postcode validation for Sweden and Turkey.
Added localization for Turkish.

onurersel committed:
Added postcode validation for Sweden and Turkey.
Added localization for Turkish.

onurersel committed:
Added credit card validation.

onurersel committed:
Added credit card validation.

onurersel committed:
‘CreditCardValidator‘ can return valid card types of a given card number.

onurersel committed:
‘CreditCardValidator‘ can return valid card types of a given card number.

Numan1617 committed:
Correct URL for project license

onurersel committed:
‘CreditCardValidator‘ can get type of card number asynchronously now, using ‘cardTypeOf‘.
Fixed fastlane tests by changing simulator sdk version.

onurersel committed:
‘CreditCardValidator‘ can get type of card number asynchronously now, using ‘cardTypeOf‘.
Fixed fastlane tests by changing simulator sdk version.

onurersel committed:
Another attempt to fix fastlane sdk error.

onurersel committed:
Another attempt to fix fastlane sdk error.

onurersel committed:
Removed device and SDK settings from Fastlane, for Travis to use default settingsi.

onurersel committed:
Removed device and SDK settings from Fastlane, for Travis to use default settingsi.

Commit:
Merge branch 'Numan1617-patch-1'

Commit:
Update change log

Commit:
Disable vertical whitespace linting

from SwiftLint rules. It overzealous and does not match project
convention of two empty lines before a MARK comment.

Commit:
Improve documentation for `validatorDelegate`

- Adds documentation for `validatorDelegate` property and
`setValidatorDelegate(_:)` method in `ValidatorTextField` and
`ValidatorTextView`.
- Callout `setValidatorDelegate(_:)` method in the README example.

Commit:
Update change log

For PR #21.

Commit:
Merge branch 'issue/19-validator-delegate-documentation'

Commit:
Fix implicit getter lint warning

- Removed `get` keyword from read-only properties as it is implicit

Commit:
Bump version to 1.1.0

Commit:
Update gem dependencies

ghost commented:
@onurersel Can you look at rebasing this PR? Now that we've merged in your last PR and bumped the version number, there's a few conflicts.

onurersel committed:
Merge branch 'master' of https://github.com/onurersel/formvalidator-swift

onurersel committed:
Merge branch 'feature/credit-card-validator' of https://github.com/onurersel/formvalidator-swift into feature/credit-card-validator

onurersel committed:
Replaces credit card violation error message for Danish, French, German, Russian, Swedish, Romanian, Portuguese, Norwegian, Polish, Korean, Japanese and Italian with English message temporarily.

onurersel commented:
Hey @aamctustwo, just rebased the branch.

ghost commented:
Thanks @onurersel . I will try and take a look at it this week.

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,102 @@
+        case .dinersClub:
+            return "3(?:0[0-5]|[68][0-9])[0-9]{11}"
+        case .discover:
+            return "6(?:011|5[0-9]{2})(?:[0-9]{12})"
```
Should this be `6(?:011|5[0-9]{2})[0-9]{12}` for Discover?

onurersel replied:
That's right, grouping is totally unnecessary.

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,102 @@
+        case .maestro:
+            return "(?:5[0678]\\d\\d|6304|6390|67\\d\\d)\\d{8,15}"
+        case .mastercard:
+            return "5[1-5][0-9]{14}"
```
Should this be `(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}` for MasterCard?

onurersel replied:
Thanks for especially this one. Had a hard time ruling out MasterCard.

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,102 @@
+        case .mastercard:
+            return "5[1-5][0-9]{14}"
+        case .visa:
+            return "^4[0-9]{12}(?:[0-9]{3})?$"
```
In all the other regex, you didn't use `^` or `$` anchors. Why did you choose to use them here?

onurersel replied:
Adding start and end tags to other regex as well 👍

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,102 @@
+            return "5[1-5][0-9]{14}"
+        case .visa:
+            return "^4[0-9]{12}(?:[0-9]{3})?$"
+        case .all:
```
Rather than having an "all" case in the enum, perhaps it would be better to create this as an OptionSet that can generate the appropriate regex and the like from the selected options?

See https://developer.apple.com/reference/swift/optionset

onurersel replied:
🤔 Interesting, I'll implement OptionSet and try to keep .all though, as a shortcut for including all available cards to condition check.

aamctustwo replied:
All can make sense in an `enum`, but I wouldn't use it as a case. This would not make logical sense in an enum:

```swift
let cards: [CreditCardType] = [.all, .visa]
```

Instead I'd make it a static property. An example would be:

```swift
public enum CreditCardType: String {
  case jcb
  case visa

  static let allTypes: [CreditCardType] = [.jcb, .visa]
}
```

I still think that OptionSet makes more sense, as you are typically going to be using this as a set of multiple values and the calculation to determine membership will be much faster with an OptionSet arithmetic rather than using `contains` on an array of enum values.

onurersel replied:
> This would not make logical sense in an enum

You're absolutely right. Updated `CreditCardType` to OptionSet on my last commit. 

Also, implemented `all` as a static value. 
```swift
static let all: CreditCardType = [.americanExpress, .dinersClub, .discover, .jcb, .maestro, .mastercard, .visa]
```

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,102 @@
+    
+    public var cardType: CreditCardType {
+        didSet {
+            self.regex = cardType.regex
```
The use of `self` is not necessary here. Convention in this framework is to only use `self` when it is required (such as in an initializer or closure).

aamctustwo commented on Sources/Validators/CreditCardValidator.swift:
```diff
@@ -0,0 +1,113 @@
+            let creditCardCondition = self.conditions[0] as! CreditCardCondition
+            let trimmedCardNumber = String(creditCardNumber.characters.filter{ $0 != " " })
+            
+            // American Express
```
Could we make this a bit more DRY by refactoring? It looks like a lot of repeated code.

aamctustwo commented on Sources/Validators/CreditCardValidator.swift:
```diff
@@ -0,0 +1,113 @@
+                }
+            }
+            
+            if self.conditions.count == 0 || !(self.conditions[0] is CreditCardCondition) {
```
A couple of points here:
1. I'd recommend using a `guard` statement here, since you're returning from the method.
2. You should use `self.conditions.isEmpty`. It's more efficient than the comparison of the count to zero and reads better.
3. You can put the `let creditCardCondition` statement in here to remove the need for the additional `is` check and it removes the force cast, which is why Travis CI is failing this build.
4. You should probably call the completion handler before calling return even if it's being called with an empty parameter. Otherwise, the developer using the framework could be left hanging waiting for the completion handler to be called.

onurersel replied:
You're right, this scope is quite a mess. Creating a separate condition for the checks seems like a more cleaner solution. 

Also not sure if should declare this function as static, since it's not dependent to the validator instance anyhow.

aamctustwo commented on Tests/Unit Tests/Conditions/CreditCardConditionTests.swift:
```diff
@@ -0,0 +1,153 @@
+        let expectedResult = true
+        
+        // Test
+        AssertCondition(conditionAll, testInput: cardNumberAmericanExpress, expectedResult: expectedResult)
```
Again, you appear to have a lot of repeated code here. Perhaps we can put some of these into an array and loop through the array with an assert on each member of the array? It'll allow expanding the number of credit cards we can validate in the future.

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,102 @@
+
+
+/// Credit card types that are supported by `CreditCardCondition`. There is a `regex` property that returns the regex for validating that credit card number.
+public enum CreditCardType : String {
```
There are a lot of SwiftLint warnings and errors on this branch (e.g. not having the colon next to the name of the type/object/key as is the case on this line). Please install SwiftLint and correct it's errors. There's only one false-positive right now which is the "weak delegate violation" in the `ValidatorControl` protocol definition.

https://github.com/realm/SwiftLint

onurersel replied:
Great tool to fight with bad habits like space before semicolons, cheers!

onurersel committed:
Fixed SwiftLint issues.
Updated credit card regular expressions.
Converted ‘CreditCardCondition‘ from enum to OptionSet.
Code convention fix on ‘CreditCardCondition‘.
Overhauled ‘cardTypeOf‘ function of ‘CreditCardValidator‘.
DRYed ‘CreditCardValidator‘.
DRYed ‘CreditCardConditionTests‘.

onurersel commented:
Great review @aamctustwo, thanks for pointing those issues out!

aamctustwo commented on Sources/Validators/CreditCardValidator.swift:
```diff
@@ -0,0 +1,75 @@
+      - creditCardNumber: Credit card number.
+      - completion: Completion callback, returns valid card types for given number.
+     */
+    public func cardTypeOf (creditCardNumber: String, completion : @escaping (CreditCardType) -> Void) {
```
🤔 I'm not sure this is a particularly clear function name. Perhaps something like

```swift
public func validCardTypes(for creditCardNumber: String, completion: @escaping (CreditCardType) -> Void) {
```

I'm open to other suggestions as well. I feel that the function name, as it stands now, is not clear as to what the function is doing.

aamctustwo replied:
It'd be good to also add a note in the header comments that this function is asynchronous. While it should be obvious given the completion handler, it's also nice to add it to the notes as well.

onurersel replied:
Yes, terrible naming here. No suggestions on this one, yours is perfectly readable.

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,156 @@
+        return check(trimmedText, withRegex: regExp)
+    }
+    
+    public func check(_ trimmedText: String, withRegex regExp: NSRegularExpression) -> Bool {
```
I think this is fine for a first implementation of a credit card condition/validator. I feel that, long-term, we should look to allow the developer to customise how they want to handle whitespace and dashes.

onurersel replied:
Agreed. There can be special cases that developer wants to check specific patterns with whitespace and dash characters like you said.

aamctustwo commented on Sources/Validators/CreditCardValidator.swift:
```diff
@@ -0,0 +1,75 @@
+        
+        DispatchQueue.global().async {
+            
+            var types: CreditCardType = []
```
There are `types` and `cardTypes` both, in different ways, storing card types. It took me a couple readings to make sense of the logic in this function. Perhaps we can be more explicit with the names of the variables? Maybe `validCardTypes` and `possibleCardTypes`? Open to suggestions here, but it would help make things clearer for the reader.

onurersel replied:
Terrible naming for variables here as well. Totally understand how you got lost.

aamctustwo commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,156 @@
+            return false
+        }
+        
+        let trimmedText = String(sourceText.characters.filter { $0 != " " })
```
Rather than using filter and only removing spaces, you should use the `trimmingCharacters(in:)` with the whitespace character set.

https://developer.apple.com/reference/swift/string/1643030-trimmingcharacters

onurersel replied:
I can't agree with this one. My intention of trimming the space characters is for processing regex with only digits from the card number. Only trimming the leading and trailing whitespace with `trimmingCharacters(in:)` is not what I'm aiming for.

But thanks for pointing this out, I updated this trimming action to leave only digit characters instead of just removing spaces.

aamctustwo commented on Sources/Validators/CreditCardValidator.swift:
```diff
@@ -0,0 +1,75 @@
+            let creditCardCondition = CreditCardCondition(cardType: .all)
+            let trimmedCardNumber = String(creditCardNumber.characters.filter { $0 != " " })
+            
+            let cardTypes: [CreditCardType] = [.americanExpress, .dinersClub, .discover, .jcb, .maestro, .mastercard, .visa]
```
I'm wondering if this should be moved into the `CreditCardType` struct as another static property, similar to the `all` option but this time providing an `all` array of options. Not sure what to call it to distinguish between them. But when we want to support new card types, all the work needs to be done in the `CreditCardType` definition and this method would not need updating.

onurersel replied:
Named it `allArray`. I think it's quite simple and understandable at this point. But any future suggestions are welcome :)

aamctustwo commented on Tests/Unit Tests/Conditions/CreditCardConditionTests.swift:
```diff
@@ -0,0 +1,76 @@
+    
+    
+    
+    // MARK: - Test Success
```
Much more DRY! I like it! :+1:

onurersel commented:
Hey @aamctustwo, great review as usual! ☺️

onurersel commented on Sources/Conditions/CreditCardCondition.swift:
```diff
@@ -0,0 +1,156 @@
+    
+    
+    var description: String {
+        switch self.rawValue {
```
Getting rid of switch here makes perfect sense.

onurersel committed:
`CreditCardType` now implements `CustomStringConvertible`.
Refactored `CreditCardType` and `CreditCardValidator`.
`CreditCardCondition` now ignores every non-digit character while checking.

ghost commented:
Closing. Will merge from PR #22 `Feature/credit card validator`. Cherry-picked the key commits onto a new branch.
